byzcoin/tx_pipeline.go

```diff
@@ -0,0 +1,394 @@
+	scsOut, sstOut, err := s.processOneTx(inState.sst, tx)
+
+	// try to create a new state
+	newState := func() *txProcessorState {
+		if err != nil {
+			return &txProcessorState{
+				inState.sst,
+				inState.scs,
+				append(inState.txs, TxResult{tx, false}),
+			}
+		}
+		return &txProcessorState{
+			sstOut,
+			append(inState.scs, scsOut...),
+			append(inState.txs, TxResult{tx, true}),
+		}
+	}()
+
+	// we're within the block size, so return one state
+	if s.GetBlockSize() > newState.size() {
+		return []*txProcessorState{newState}, nil
+	}
+
+	// if the new state is too big, we split it
+	newStates := []*txProcessorState{inState.copy()}
+	if err != nil {
+		newStates = append(newStates, &txProcessorState{
+			inState.sst,
+			inState.scs,
+			[]TxResult{{tx, false}},
+		})
+	} else {
+		newStates = append(newStates, &txProcessorState{
+			sstOut,
+			scsOut,
+			[]TxResult{{tx, true}},
+		})
+	}
+	return newStates, nil
+}
+
+// ProposeBlock basically calls s.createNewBlock which might block. There is
+// nothing we can do about it other than waiting for the timeout.
+func (s *defaultTxProcessor) ProposeBlock(state *txProcessorState) error {
+	config, err := LoadConfigFromTrie(state.sst)
+	if err != nil {
+		return err
+	}
+	_, err = s.createNewBlock(s.scID, &config.Roster, state.txs)
+	return err
+}
+
+func (s *defaultTxProcessor) GetInterval() time.Duration {
+	bcConfig, err := s.LoadConfig(s.scID)
+	if err != nil {
+		log.Error(s.ServerIdentity(), "couldn't get configuration - this is bad and probably "+
+			"a problem with the database! "+err.Error())
+		return defaultInterval
+	}
+	return bcConfig.BlockInterval
+}
+
+func (s *defaultTxProcessor) GetLatestGoodState() *txProcessorState {
+	st, err := s.getStateTrie(s.scID)
+	if err != nil {
+		// A good state must exist because we're working on a known
+		// skipchain. If there is an error, then the database must've
+		// failed, so there is nothing we can do to recover so we
+		// panic.
+		panic(fmt.Sprintf("failed to get a good state: %v", err))
+	}
+	return &txProcessorState{
+		sst: st.MakeStagingStateTrie(),
+	}
+}
+
+func (s *defaultTxProcessor) GetBlockSize() int {
+	bcConfig, err := s.LoadConfig(s.scID)
+	if err != nil {
+		log.Error(s.ServerIdentity(), "couldn't get configuration - this is bad and probably "+
+			"a problem with the database! "+err.Error())
+		return defaultMaxBlockSize
+	}
+	return bcConfig.MaxBlockSize
+}
+
+func (s *defaultTxProcessor) Stop() {
+	close(s.stopCollect)
+}
+
+type txPipeline struct {
+	wg        sync.WaitGroup
+	processor txProcessor
+}
+
+func (p *txPipeline) start(initialState *txProcessorState, stopChan chan bool) {
+	ctxChan, stopChan1 := p.collectTx()
+	stopChan2 := p.processTxs(ctxChan, initialState)
+
+	// wait for signal and then stop
+	<-stopChan
+	close(stopChan1)
+	close(stopChan2)
+	p.processor.Stop()
+	p.wg.Wait()
+}
+
+func (p *txPipeline) collectTx() (<-chan ClientTransaction, chan<- bool) {
+	stopChan := make(chan bool, 1)
```

Comment:
Same as the other stop channel, is the buffering necessary ?

Reply:
nope